rust_snuba/src/processors/replays.rs

```diff
@@ -0,0 +1,573 @@
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use uuid::Uuid;
+
+use crate::types::{InsertBatch, KafkaMessageMetadata, RowData};
+
+pub fn process_message(
+    payload: KafkaPayload,
+    metadata: KafkaMessageMetadata,
+) -> anyhow::Result<InsertBatch> {
+    let payload_bytes = payload.payload().context("Expected payload")?;
+
+    let replay_message: ReplayMessage = serde_json::from_slice(payload_bytes)?;
+    let replay_payload = serde_json::from_slice(&replay_message.payload)?;
+
+    let output = match replay_payload {
+        ReplayPayload::ClickEvent(event) => {
+            let replay_rows = event.clicks.into_iter().map(|click| ReplayRow {
+                click_alt: click.alt,
+                click_aria_label: click.aria_label,
+                click_class: click.class,
+                click_id: click.id,
+                click_is_dead: click.is_dead,
+                click_is_rage: click.is_rage,
+                click_node_id: click.node_id,
+                click_role: click.role,
+                click_tag: click.tag,
+                click_testid: click.testid,
+                click_text: click.text,
+                click_title: click.title,
+                event_hash: click.event_hash,
+                offset: metadata.offset,
+                partition: metadata.partition,
+                project_id: replay_message.project_id,
+                replay_id: replay_message.replay_id,
+                retention_days: replay_message.retention_days,
+                timestamp: click.timestamp,
+                ..Default::default()
+            });
+
+            let mut rows: Vec<Vec<u8>> = Vec::with_capacity(replay_rows.len());
+
+            for row in replay_rows {
+                rows.push(serde_json::to_vec(&row)?);
+            }
+
+            rows
+        }
+        ReplayPayload::Event(event) => {
+            let event_hash = match (event.event_hash, event.segment_id) {
+                (None, None) => Uuid::new_v4(),
+                (None, Some(segment_id)) => {
+                    Uuid::from_slice(md5::compute(segment_id.to_string().as_bytes()).as_slice())?
+                }
+                (Some(event_hash), _) => event_hash,
+            };
+
+            let (ip_address_v4, ip_address_v6) = match event.user.ip_address {
+                None => (None, None),
+                Some(IpAddr::V4(ip)) => (Some(ip), None),
+                Some(IpAddr::V6(ip)) => (None, Some(ip)),
+            };
+
+            let user = event.user.user_id.clone();
+
+            let tags_key: Vec<String> = event
+                .tags
+                .clone()
+                .iter()
+                .map(|v| v.first().unwrap_or(&String::new()).to_owned())
+                .collect();
+
+            let tags_value: Vec<String> = event
+                .tags
+                .clone()
+                .iter()
+                .map(|v| v.get(1).unwrap_or(&String::new()).to_owned())
+                .collect();
```

Comment:
```suggestion
            let tags_key: Vec<_> = event
                .tags
                .iter()
                .map(|v| v.first().unwrap_or_default().to_owned())
                .collect();

            let tags_value: Vec<_> = event
                .tags
                .into_iter()
                .map(|v| v.get(1).unwrap_or_default())
                .collect();
```

Previously, you were creating two clones of the entire `Vec`, just to clone its contents again.
Now this should avoid at least one clone.

If you went with `Vec<(String,String)>` as I suggested, you can also use `.into_iter().unzip()` directly to split it into two `Vec`s.